discovery-provider/src/tasks/cache_user_balance.py

```diff
@@ -85,26 +90,35 @@ def refresh_user_ids(redis, db, token_contract, eth_web3):
         for user in user_query:
             user_id, user_wallet, associated_wallet = user
             if not user_id in user_id_wallets:
-                user_id_wallets[user_id] = [user_wallet]
+                user_id_wallets[user_id] = {"owner_wallet": user_wallet}
             if associated_wallet:
-                user_id_wallets[user_id].append(associated_wallet)
+                if not "associated_wallets" in user_id_wallets[user_id]:
+                    user_id_wallets[user_id]["associated_wallets"] = [associated_wallet]
+                else:
+                    user_id_wallets[user_id]["associated_wallets"].append(associated_wallet)
 
         logger.info(f"cache_user_balance.py | fetching for {len(user_query)} users: {needs_refresh_map.keys()}")
 
         # Fetch balances
         for user_id, wallets in user_id_wallets.items():
             try:
-                total = 0
-                for wallet in wallets:
-                    wallet = eth_web3.toChecksumAddress(wallet)
-
-                    # get balance
-                    balance = token_contract.functions.balanceOf(wallet).call()
-                    total += balance
+                owner_wallet = wallets["owner_wallet"]
+                owner_wallet = eth_web3.toChecksumAddress(owner_wallet)
+                owner_wallet_balance = token_contract.functions.balanceOf(owner_wallet).call()
+                associated_balance = 0
+
+                if "associated_wallets" in wallets:
+                    for wallet in wallets["associated_wallets"]:
+                        wallet = eth_web3.toChecksumAddress(wallet)
+                        balance = token_contract.functions.balanceOf(wallet).call()
+                        delegation_balance = delegate_manager_contract.functions.getTotalDelegatorStake(wallet).call()
+                        stake_balance = staking_contract.functions.totalStakedFor(wallet).call()
+                        associated_balance += balance + delegation_balance + stake_balance
```

Comment:
So these are definitely safe to just add together, yeah?

Reply:
yeah!